app/products/agents/components/tool_approval_set/index.tsx

```diff
@@ -0,0 +1,207 @@
+            const updatedCount = Object.keys(updated).length;
+            const prevCount = Object.keys(prev).length;
+            return updatedCount === prevCount ? prev : updated;
+        });
+    }, [toolCalls]);
+
+    const handleToolDecision = async (toolId: string, approved: boolean) => {
+        if (isSubmitting) {
+            return;
+        }
+
+        const updatedDecisions = {
+            ...toolDecisions,
+            [toolId]: approved,
+        };
+        setToolDecisions(updatedDecisions);
+
+        // Check if there are still undecided tools
+        const hasUndecided = toolCalls.some((tool) => {
+            return !(tool.id in updatedDecisions) || updatedDecisions[tool.id] === null;
+        });
+
+        if (hasUndecided) {
+            // If there are still undecided tools, do not submit yet
+            return;
+        }
+
+        // Submit when all tools are decided
+        const approvedToolIds = Object.entries(updatedDecisions).
+            filter(([, isApproved]) => isApproved).
+            map(([id]) => id);
+
+        setIsSubmitting(true);
+        const {error} = await submitToolApproval(serverUrl, postId, approvedToolIds);
+
+        if (error) {
+            // Reset on error so user can try again
+            setIsSubmitting(false);
+        } else {
+            // Backend will execute tools and update the post via POST_EDITED event
+            // The component will receive updated toolCalls prop and useEffect will clear local decisions
+            setIsSubmitting(false);
+        }
+    };
+
+    const toggleCollapse = (toolId: string, defaultExpanded: boolean) => {
+        setExpandedTools((prev) => ({
+            ...prev,
+            [toolId]: !(prev[toolId] ?? defaultExpanded),
+        }));
+    };
+
+    if (toolCalls.length === 0) {
+        return null;
+    }
+
+    // Get pending tool calls
+    const pendingToolCalls = toolCalls.filter((call) => call.status === ToolCallStatus.Pending);
+
+    // Get processed tool calls
+    const processedToolCalls = toolCalls.filter((call) => call.status !== ToolCallStatus.Pending);
+
+    // Calculate how many pending tools haven't been decided yet
+    const undecidedCount = pendingToolCalls.filter(
+        (tool) => !(tool.id in toolDecisions),
+    ).length;
+
+    // Helper to compute if a tool should be collapsed
+    const isToolCollapsed = (tool: ToolCall) => {
+        const defaultExpanded = tool.status === ToolCallStatus.Pending;
+        return !(expandedTools[tool.id] ?? defaultExpanded);
+    };
+
+    return (
+        <View style={styles.container}>
+            {pendingToolCalls.map((tool) => (
+                <ToolCard
+                    key={tool.id}
+                    tool={tool}
+                    isCollapsed={isToolCollapsed(tool)}
+                    isProcessing={isSubmitting}
+                    localDecision={toolDecisions[tool.id]}
+                    onToggleCollapse={() => toggleCollapse(tool.id, true)}
```

Comment:
I'm not a fan of inline prop functions, we can avoid these children from re-rendering if handled differently

Reply:
Fixed! Refactored ToolCard to accept toolId-based callbacks. Parent now passes memoized handleApprove, handleReject, and toggleCollapse callbacks wrapped in useCallback. ToolCard internally calls onApprove?.(tool.id) etc.